core/base/validation.hpp

```diff
@@ -0,0 +1,183 @@
+// SPDX-FileCopyrightText: 2017 - 2025 The Ginkgo authors
+//
+// SPDX-License-Identifier: BSD-3-Clause
+
+#ifndef GKO_CORE_BASE_VALIDATION_HPP_
+#define GKO_CORE_BASE_VALIDATION_HPP_
+
+
+#include <cmath>
+#include <unordered_set>
+
+#include <ginkgo/core/base/array.hpp>
+#include <ginkgo/core/base/temporary_clone.hpp>
+
+
+namespace gko {
+namespace validation {
+
+
+#define GKO_VALIDATE(_expression, _message)                                 \
+    if (!(_expression)) {                                                   \
+        throw gko::InvalidData(__FILE__, __LINE__, typeid(decltype(*this)), \
+                               _message);                                   \
+    }
+
+
+template <typename ValueType>
+bool is_finite_scalar(const ValueType& value)
+{
+    return std::isfinite(static_cast<double>(value));
+}
+
+template <typename ValueType>
+bool is_finite_scalar(const std::complex<ValueType>& value)
+{
+    return is_finite_scalar(value.real()) && is_finite_scalar(value.imag());
+}
+
+
+template <typename IndexType>
+bool is_sorted(const gko::array<IndexType>& row_ptrs)
+{
+    const auto host_row_ptrs =
+        make_temporary_clone(row_ptrs.get_executor()->get_master(), &row_ptrs);
+    return std::is_sorted(
+        host_row_ptrs->get_const_data(),
+        host_row_ptrs->get_const_data() + host_row_ptrs->get_size());
+}
+
+
+template <typename IndexType>
+bool is_within_bounds(const gko::array<IndexType>& col_idxs,
+                      const std::int64_t& lower_bound,
+                      const size_t& upper_bound)
```

Comment:
I think those two types should be consistent with the value type of the array, at worst both should be int64, maybe even
```suggestion
                      IndexType lower_bound,
                      IndexType upper_bound)
```
Additionally, I'm not sure if we even need the lower_bound parameter, since it's always set to 0